src/Serialization/main.jl

```diff
@@ -204,6 +205,57 @@ function load_internal_generic(s::DeserializerState, ::Type{T}, dict::Dict) wher
     return T(fields...)
 end
 
+################################################################################
+# Include internal saves for specific types
+
+include("basic_types.jl")
+include("containers.jl")
+include("PolyhedralGeometry.jl")
+include("Combinatorics.jl")
+include("Fields.jl")
+include("ToricGeometry.jl")
+include("Rings.jl")
+include("polymake.jl")
+include("TropicalGeometry.jl")
+include("QuadForm.jl")
+
+################################################################################
+# Include upgrade scripts
+
+include("upgrades/main.jl")
+
+################################################################################
+# Loading with upgrade checks on dict
+
+# Finds the first version where an upgrade can be applied and then incrementally
+# upgrades to the current version
+function upgrade(dict::Dict{Symbol, Any}, dict_version::VersionNumber)
+    upgraded_dict = dict
+    for upgrade_script in upgrade_scripts
+        script_version = version(upgrade_script)
+
+        if dict_version < script_version
+            println("upgrading serialized data to version $script_version")
+            s = DeserializerState()
+
+            upgraded_dict = upgrade_script(s, upgraded_dict)
+        end
+    end
+
+    upgraded_dict[:_ns] = get_version_info()
+    
+    return upgraded_dict
+end
+
+function get_file_version(dict::Dict{Symbol, Any})
```

Comment:
Off-topic / not for this PR but for future consideration: perhaps it is time to add a dedicated type for serialized data. Or rather, two: `abstract type AbstractSerializedData` and then 
```
struct JSONSerializedData <: AbstractSerializedData
  data::Dict{Symbol, Any}
end
```
Because I still want us to eventually add a serialization format that does *not* construct this intermediate tree representation, see #1507, and that means we need clean interfaces.

But as I said, that's not for this PR, but perhaps a follow-up PR.